### PR TITLE
Create PipelineCacheEvent to report changes to pipelines in PipelineCache.

### LINKED ISSUE #9616: Send an Event whenever a pipeline is (re)loaded by the PipelineCache.
## What problem does this solve or what need does it fill?

With compute shaders, some shaders may require multiple steps (e.g., initialization of textures, then normal update). Hot-reloading shaders can change a pipeline at runtime, but it is non-trivial to react to that (e.g., resetting a RenderGraph node to recompute initial states).

## What solution would you like?

Whenever the PipelineCache changes the state of a pipeline, send a `PipelineCacheEvent` (or queue up an event and send it later).

This allows not only reacting to hot-reloaded pipelines, but also reacting to newly-loaded pipelines! Instead of checking whether pipelines are loaded in the `render_graph::Node::update` and updating state, users can also change the state of a `render_graph::Node` from a regular system.

## What alternative(s) have you considered?

* It is possible to `Extract<EventReader<AssetEvent<Shader>>>` and track which shader a pipeline is using. This requires a lot of bookkeeping.
* It may be possible to add a system between the `PipelineCache::extract_shaders` system and the `PipelineCache::process_pipeline_queue_system` system to detect a difference between them. This requires a crazy amount of bookkeeping.

### PR DESCRIPTION
**Pull request by andriyDev**
# Objective

Previously, the only way to tell if a pipeline had been loaded was by polling for the CachedPipelineState, and there was no way to know if a pipeline had been reloaded (for example because of hot-reloading a shader). Now, youcan listen for PipelineCacheEvents to know when a Pipeline has been created or invalidated (or even failed to create)!

Fixes #9616.

## Solution

- Describe the solution used to achieve the objective above.

---

## Changelog

- Added PipelineCacheEvent to notify users of changes to pipelines stored in the PipelineCache.

## Extra notes

I tested this on my personal project, and was able to reset my `render_graph::Node` to its "loading" state whenever I updated my shader through the filewatcher feature! Now playing with compute shaders is a ton easier!

### CODE REVIEW

File: crates/bevy_render/src/render_resource/pipeline_cache.rs [hunk 1]
```diff
@@ -79,6 +84,27 @@ impl CachedComputePipelineId {
...
 
+/// Index of a cached pipeline of any kind in a [`PipelineCache`]. This is essentially a union of
+/// all pipeline ID types.
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub enum CachedAnyPipelineId {
+    RenderPipeline(CachedRenderPipelineId),
+    ComputePipeline(CachedComputePipelineId),
+}
+
+impl CachedAnyPipelineId {
```
Comment (notmd) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:95]: I think the name `AnyCachedPipelineId` makes more sense here. I'm not a native speaker btw.

Comment (andriyDev) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs]: Done. Honestly I was having a hard time deciding between the two :P I initially decided to go with `CachedAnyPipelineId` to match the `CachedComputePipelineId` and `CachedRenderPipelineId`, but I am personally fine with `AnyCachedPipelineId`.

**Review by JMS55 — COMMENTED** [review]
Code looks fine to me, and I don't think I have any objections to adding the feature. 

Curious to know what the use case for this is though. Why does your node care about when the pipeline gets reloaded?

**Comment by andriyDev** [comment]
> Code looks fine to me, and I don't think I have any objections to adding the feature.
> 
> Curious to know what the use case for this is though. Why does your node care about when the pipeline gets reloaded?

TL;DR graph nodes with "setup" stages may need to be reset when a shader reloads (or more generally a pipeline reloads) so the setup stages reruns. Not strictly necessary, but quite nice!

**Comment by JMS55** [comment]
I feel like this is easier to solve by just storing the last pipeline id in the node, and when you get the pipeline from the cache in your node, just compare it to the last stored id. Like I said I'm not against this feature in general though.

Would like to hear what kind of setup your doing in more concrete terms though. I'm curious what kind of data or whatever you need to reset.

**Comment by andriyDev** [comment]
I guess "easier" is in the eye of the beholder haha. I was actually able to make this work, so that seems to be a valid workaround, thank you! It does mean having to keep track of the `PipelineId` which are completely distinct from the `CachedPipelineId` (it would be nice to note in the `CachedPipelineId` docs that the PipelineId can change underneath it). I was able to do that with this function (including just for interest):

```rust
fn update_pipeline(
  cached_pipeline_id: CachedComputePipelineId,
  last_pipeline_id: &mut Option<ComputePipelineId>,
  pipeline_cache: &PipelineCache,
) -> bool {
  let mut require_reset = false;

  match pipeline_cache.get_compute_pipeline(cached_pipeline_id) {
    None => {
      *last_pipeline_id = None;
      require_reset = true;
    }
    Some(pipeline) => match *last_pipeline_id {
      None => {
        *last_pipeline_id = Some(pipeline.id());
      }
      Some(current_pipeline) => {
        if pipeline.id() != current_pipeline {
          *last_pipeline_id = Some(pipeline.id());
          require_reset = true;
        }
      }
    },
  }

  require_reset
}
```

---

As for my concrete case: I am trying to make an ocean "simulation" compute shader. This involves roughly 3 steps (simplified because I'm not fully done): 1) generate random noise on startup (on the CPU), 2) use the noise to determine the amplitudes of waves on startup (or whenever params change), 3) use FFTs to turn "frequency-space" wave amplitudes into the actual heights of the waves every frame.

I have a render graph node:
```rust
enum OceanSimulationNode {
  Loading,
  InitWaveAmplitudes,
  UpdateWaterHeight,
}
```

In my `render_graph::Node::update` (`is_ready` just checks if all `CachedComputePipelineId`s have an Ok status in the `PipelineCache`):
```rust
match self {
  Self::Loading => {
    if self.is_ready(world) {
      *self = OceanSimulationNode::InitWaveAmplitudes;
    }
  }
  Self::InitWaveAmplitudes => {
    *self = OceanSimulationNode::UpdateWaterHeight;
  }
  Self::UpdateWaterHeight => {}
};
```

With hot-reloading shaders, I can change the FFT shader and see changes immediately. However, if I change the "step 2" shader, the `OceanSimulationNode` is already in the `UpdateWaterHeight` state, and so the pipeline never runs.

---

But as I said above, this can be worked around with keeping track of the `PipelineId` in additional to the `CachedPipelineId`. Personally I like the event scheme since A) it involves less bookkeeping by users (e.g., me), and B) it's more inline with Bevy as a whole I'd say? One disadvantage is this kinda requires two "flushes" as I have in the code - one for when shader events are extracted, and one after the render system, so it is less nice to render in response to a pipeline being ready (there will likely be a 1 frame delay).

I'm not sure which is the better solution! What are your thoughts?

**Comment by andriyDev** [comment]
I'm gonna just close this because it's been inactive and there's a fine workaround. I'll also close the issue and add the workaround there.
